File: ts/src/shapes.ts
import { z } from "zod";
import { FacetError, InvariantError } from "./errors.js";
import type { Node } from "./node.js";
import type { KindSpec, Registry, ShapeSpec } from "./registry.js";
import { RelationSchema } from "./relations.js";

export const MEMBERSHIP = "membership";
export const EDGES = "edges";
export const ORDER = "order";
export const KEYS = "keys";

export const MembershipSchema = z.object({ members: z.array(z.string()).default([]) });
export const EdgesSchema = z.object({ edges: z.array(RelationSchema).default([]) });
export const OrderSchema = z.object({ order: z.array(z.string()).default([]) });
export const KeysSchema = z.object({ keys: z.record(z.string()).default({}) });

export type Membership = z.infer<typeof MembershipSchema>;
export type Edges = z.infer<typeof EdgesSchema>;
export type Order = z.infer<typeof OrderSchema>;
export type Keys = z.infer<typeof KeysSchema>;

function load<S extends z.ZodTypeAny>(node: Node, name: string, schema: S): z.output<S> {
  const raw = node.facets[name];
  if (raw === undefined) throw new FacetError(`${node.id}: missing '${name}' facet`);
  const result = schema.safeParse(raw);
  if (!result.success) {
    throw new FacetError(
      `${node.id}: invalid '${name}' facet: ${result.error.issues.map((i) => i.message).join("; ")}`,
    );
  }
  return result.data as z.output<S>;
}

export function membershipOf(node: Node): Membership {
  return load(node, MEMBERSHIP, MembershipSchema);
}
export function edgesOf(node: Node): Edges {
  return load(node, EDGES, EdgesSchema);
}
export function orderOf(node: Node): Order {
  return load(node, ORDER, OrderSchema);
}
export function keysOf(node: Node): Keys {
  return load(node, KEYS, KeysSchema);
}

export function requireUniqueMembers(node: Node): void {
  const members = membershipOf(node).members;
  if (members.length !== new Set(members).size) {
    throw new InvariantError(`${node.id}: members must be unique`);
  }
}

export function requireEdgeEndpointsAreMembers(node: Node): void {
  const members = new Set(membershipOf(node).members);
  for (const e of edgesOf(node).edges) {
    if (!members.has(e.source) || !members.has(e.target)) {
      throw new InvariantError(`${node.id}: edge endpoints must be members`);
    }
  }
}

export function requireOrderIsPermutation(node: Node): void {
  const members = membershipOf(node).members;
  const order = orderOf(node).order;
  const memberSet = new Set(members);
  const orderSet = new Set(order);
  const sameSet = memberSet.size === orderSet.size && [...memberSet].every((m) => orderSet.has(m));
  if (order.length !== members.length || !sameSet) {
    throw new InvariantError(`${node.id}: order must be a permutation of members`);
  }
}

export function requireKeyValuesAreMembers(node: Node): void {
  const members = new Set(membershipOf(node).members);
  for (const value of Object.values(keysOf(node).keys)) {
    if (!members.has(value)) {
      throw new InvariantError(`${node.id}: key values must be members`);
    }
  }
}

export function requireAcyclic(node: Node): void {
  const adj = new Map<string, string[]>();
  for (const e of edgesOf(node).edges) {
    const list = adj.get(e.source) ?? [];
    list.push(e.target);
    adj.set(e.source, list);
  }
  const visiting = new Set<string>();
  const done = new Set<string>();
  const walk = (n: string): void => {
    if (visiting.has(n)) throw new InvariantError(`${node.id}: cycle detected at ${n}`);
    if (done.has(n)) return;
    visiting.add(n);
    for (const nxt of adj.get(n) ?? []) walk(nxt);
    visiting.delete(n);
    done.add(n);
  };
  for (const start of [...adj.keys()]) walk(start);
}

export function requireSingleParent(node: Node): void {
  const parents = new Map<string, number>();
  for (const e of edgesOf(node).edges) {
    parents.set(e.target, (parents.get(e.target) ?? 0) + 1);
  }
  const over = [...parents.entries()]
    .filter(([, c]) => c > 1)
    .map(([t]) => t)
    .sort();
  if (over.length > 0) {
    throw new InvariantError(`${node.id}: nodes with multiple parents: ${JSON.stringify(over)}`);
  }
}

export function registerBuiltinShapes(reg: Registry): void {
  const shapes: ShapeSpec[] = [
    { name: "set", requiredFacets: new Set([MEMBERSHIP]), invariants: [requireUniqueMembers] },
    {
      name: "list",
      requiredFacets: new Set([MEMBERSHIP, ORDER]),
      invariants: [requireUniqueMembers, requireOrderIsPermutation],
    },
    {
      name: "dict",
      requiredFacets: new Set([MEMBERSHIP, KEYS]),
      invariants: [requireUniqueMembers, requireKeyValuesAreMembers],
    },
    {
      name: "graph",
      requiredFacets: new Set([MEMBERSHIP, EDGES]),
      invariants: [requireUniqueMembers, requireEdgeEndpointsAreMembers],
    },
    {
      name: "dag",
      requiredFacets: new Set([MEMBERSHIP, EDGES]),
      invariants: [requireUniqueMembers, requireEdgeEndpointsAreMembers, requireAcyclic],
    },
    {
      name: "tree",
      requiredFacets: new Set([MEMBERSHIP, EDGES]),
      invariants: [requireUniqueMembers, requireEdgeEndpointsAreMembers, requireAcyclic, requireSingleParent],
    },
  ];
  for (const shape of shapes) reg.registerShape(shape);
  for (const name of ["set", "list", "dict", "graph", "dag", "tree"]) {
    const spec: KindSpec = { name, shape: name };
    reg.register(spec);
  }
}

File: python/src/nodes/core/shapes.py
from __future__ import annotations

from typing import TypeVar

from pydantic import BaseModel, Field
from pydantic import ValidationError as PydanticValidationError

from nodes.core.errors import FacetError, InvariantError
from nodes.core.node import Node
from nodes.core.registry import KindSpec, Registry, ShapeSpec
from nodes.core.relations import Relation

_M = TypeVar("_M", bound=BaseModel)

MEMBERSHIP = "membership"
EDGES = "edges"
ORDER = "order"
KEYS = "keys"


class Membership(BaseModel):
    members: list[str] = Field(default_factory=list)


class Edges(BaseModel):
    edges: list[Relation] = Field(default_factory=list)


class Order(BaseModel):
    order: list[str] = Field(default_factory=list)


class Keys(BaseModel):
    keys: dict[str, str] = Field(default_factory=dict)


def _load(node: Node, name: str, model: type[_M]) -> _M:
    raw = node.facets.get(name)
    if raw is None:
        raise FacetError(f"{node.id}: missing {name!r} facet")
    try:
        return model.model_validate(raw)
    except PydanticValidationError as exc:
        raise FacetError(f"{node.id}: invalid {name!r} facet: {exc}") from exc


def membership_of(node: Node) -> Membership:
    return _load(node, MEMBERSHIP, Membership)


def edges_of(node: Node) -> Edges:
    return _load(node, EDGES, Edges)


def order_of(node: Node) -> Order:
    return _load(node, ORDER, Order)


def keys_of(node: Node) -> Keys:
    return _load(node, KEYS, Keys)


def require_unique_members(node: Node) -> None:
    members = membership_of(node).members
    if len(members) != len(set(members)):
        raise InvariantError(f"{node.id}: members must be unique")


def require_edge_endpoints_are_members(node: Node) -> None:
    members = set(membership_of(node).members)
    for e in edges_of(node).edges:
        if e.source not in members or e.target not in members:
            raise InvariantError(f"{node.id}: edge endpoints must be members")


def require_order_is_permutation(node: Node) -> None:
    members = membership_of(node).members
    order = order_of(node).order
    if len(order) != len(members) or set(order) != set(members):
        raise InvariantError(f"{node.id}: order must be a permutation of members")


def require_key_values_are_members(node: Node) -> None:
    members = set(membership_of(node).members)
    for value in keys_of(node).keys.values():
        if value not in members:
            raise InvariantError(f"{node.id}: key values must be members")


def require_acyclic(node: Node) -> None:
    adj: dict[str, list[str]] = {}
    for e in edges_of(node).edges:
        adj.setdefault(e.source, []).append(e.target)
    visiting, done = set(), set()

    def walk(n: str) -> None:
        if n in visiting:
            raise InvariantError(f"{node.id}: cycle detected at {n}")
        if n in done:
            return
        visiting.add(n)
        for nxt in adj.get(n, []):
            walk(nxt)
        visiting.discard(n)
        done.add(n)

    for start in list(adj):
        walk(start)


def require_single_parent(node: Node) -> None:
    parents: dict[str, int] = {}
    for e in edges_of(node).edges:
        parents[e.target] = parents.get(e.target, 0) + 1
    over = sorted(t for t, c in parents.items() if c > 1)
    if over:
        raise InvariantError(f"{node.id}: nodes with multiple parents: {over}")


def register_builtin_shapes(reg: Registry) -> None:
    reg.register_shape(ShapeSpec(name="set", required_facets={MEMBERSHIP},
                                 invariants=[require_unique_members]))
    reg.register_shape(ShapeSpec(name="list", required_facets={MEMBERSHIP, ORDER},
                                 invariants=[require_unique_members, require_order_is_permutation]))
    reg.register_shape(ShapeSpec(name="dict", required_facets={MEMBERSHIP, KEYS},
                                 invariants=[require_unique_members, require_key_values_are_members]))
    reg.register_shape(ShapeSpec(name="graph", required_facets={MEMBERSHIP, EDGES},
                                 invariants=[require_unique_members, require_edge_endpoints_are_members]))
    reg.register_shape(ShapeSpec(name="dag", required_facets={MEMBERSHIP, EDGES},
                                 invariants=[require_unique_members, require_edge_endpoints_are_members,
                                             require_acyclic]))
    reg.register_shape(ShapeSpec(name="tree", required_facets={MEMBERSHIP, EDGES},
                                 invariants=[require_unique_members, require_edge_endpoints_are_members,
                                             require_acyclic, require_single_parent]))
    for name in ("set", "list", "dict", "graph", "dag", "tree"):
        reg.register(KindSpec(name=name, shape=name))
